### PR TITLE
NEP-35 duck array update

### PR DESCRIPTION
**Pull request by jrbourbeau**
This tests out adding an `__array_function__` condition to our `is_arraylike` check (xref https://github.com/dask/dask/issues/7263#issuecomment-786088835). Let's see if CI passes 

cc @pentschev @jakirkham 

- [ ] Closes https://github.com/dask/dask/issues/7263
- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @shoyer

**Comment by jrbourbeau** [comment]
Just to summarize the test failures we're seeing:

- The mindeps build is failing because we support versions of NumPy that predate the addition of `__array_function__` which causes `is_arraylike` to return `False` on things like NumPy arrays
- There is one test, `dask/dataframe/tests/test_dataframe.py::test_dask_dataframe_holds_scipy_sparse_containers` which is failing because [`scipy.sparse` arrays don't implement `__array_function__`](https://github.com/scipy/scipy/blob/5f4c4d802e5a56708d86909af6e5685cd95e6e66/scipy/sparse/base.py#L814-L832). FWIW I tracked this test back to https://github.com/dask/dask/pull/3738 where it looks like we don't fully support `scipy.sparse` arrays inside Dask arrays, but it is useful to partially support for some Dask-ML work (hand chunks off to `dask.delayed` or `da.map_blocks` workflows)

**Comment by jakirkham** [comment]
We also support CuPy sparse matrices for similar reasons, which likely have similar issues

**Comment by pentschev** [comment]
Thanks @jrbourbeau , you beat me to it on working this PR. 😄 



> * The mindeps build is failing because we support versions of NumPy that predate the addition of `__array_function__` which causes `is_arraylike` to return `False` on things like NumPy arrays

For this, we'll also want to check for https://github.com/dask/dask/blob/5273e73ced76862e4d73f2ac4e7373c59540c3f0/dask/array/utils.py#L524 .

> * There is one test, `dask/dataframe/tests/test_dataframe.py::test_dask_dataframe_holds_scipy_sparse_containers` which is failing because [`scipy.sparse` arrays don't implement `__array_function__`](https://github.com/scipy/scipy/blob/5f4c4d802e5a56708d86909af6e5685cd95e6e66/scipy/sparse/base.py#L814-L832). FWIW I tracked this test back to #3738 where it looks like we don't fully support `scipy.sparse` arrays inside Dask arrays, but it is useful to partially support for some Dask-ML work (hand chunks off to `dask.delayed` or `da.map_blocks` workflows)

Seems like we'll need to special-case this as well in `is_arraylike`, it doesn't look like we have a much better alternative.

File: dask/utils.py [hunk 1]
```diff
@@ -1124,12 +1124,23 @@ def is_arraylike(x):
     False
     """
     from .base import is_dask_collection
+    from .array.utils import IS_NEP18_ACTIVE
+
+    # Fallback to True for cases where NEP-18 isn't active (e.g. older versions of NumPy)
+    implements_array_fuction = (
+        hasattr(x, "__array_function__") if IS_NEP18_ACTIVE else True
+    )
```
Comment (jrbourbeau) [inline dask/utils.py:1132]: This doesn't totally work as we can't distinguish between "duck" and "input" arrays for all NumPy versions we support just using `__array_function__`. Thoughts on using `hasattr(x, "__array_function__") or hasattr(x, "__array_ufunc__")` instead?

File: dask/utils.py [hunk 2]
```diff
@@ -1124,12 +1124,23 @@ def is_arraylike(x):
...
```
Comment (jakirkham) [inline dask/utils.py:1133]:
```suggestion

    x_typename = typename(type(x))
```

File: dask/utils.py [hunk 3]
```diff
@@ -1124,12 +1124,23 @@ def is_arraylike(x):
...
     return bool(
         hasattr(x, "shape")
         and isinstance(x.shape, tuple)
         and hasattr(x, "dtype")
         and not any(is_dask_collection(n) for n in x.shape)
+        # We special case scipy.sparse arrays as having partial support for them is
+        # useful in scenerios where we mostly call `map_partitions` or `map_blocks`
+        # with scikit-learn functions on dask arrays and dask dataframes.
+        # https://github.com/dask/dask/pull/3738
+        and (implements_array_fuction or "scipy.sparse" in typename(type(x)))
```
Comment (jakirkham) [inline dask/utils.py:1143]:
```suggestion
        and (implements_array_fuction or "scipy.sparse" in x_typename or "cupyx.scipy.sparse" in x_typename)
```

**Review by jakirkham — COMMENTED** [review]
Some suggestions for CuPy sparse matrices

File: dask/utils.py (hunk 3 above)
Comment (jrbourbeau) [reply dask/utils.py:1143]: My thinking was that `"scipy.sparse"` would cover both the actual `scipy.sparse` case as well as the CuPy sparse case. Is having two checks against `x_typename` just for code clarity?

Comment (jakirkham) [reply dask/utils.py:1143]: Ah good point. Wasn't thinking about it. Yeah that's fine, but maybe we could include that in the comment above?

File: dask/utils.py (hunk 1 above)
Comment (pentschev) [reply dask/utils.py:1132]: Seems reasonable to me, unless there are non-duck arrays that also implement `__array_ufunc__`, I'm not sure if that's the case but it's worth a try.

File: dask/utils.py (hunk 3 above)
Comment (jrbourbeau) [reply dask/utils.py:1143]: 👍 just updated the comment to explicitly mentioned `cupyx.scipy.sparse` arrays

Comment (jakirkham) [reply dask/utils.py:1143]: Great thanks! Looks good 😄

File: dask/utils.py [hunk 4]
```diff
@@ -1125,11 +1125,18 @@ def is_arraylike(x):
     """
     from .base import is_dask_collection
 
+    is_duck_array = hasattr(x, "__array_function__") or hasattr(x, "__array_ufunc__")
```
Comment (jakirkham) [inline dask/utils.py:1128]: @pentschev do you have thoughts on this formulation?

File: dask/utils.py (hunk 1 above)
Comment (jakirkham) [reply dask/utils.py]: This has been updated below. Noted in this review comment ( https://github.com/dask/dask/pull/7321#discussion_r587917035 ). Would be good to hear what people think of this formulation

File: dask/utils.py (hunk 4 above)
Comment (jrbourbeau) [reply dask/utils.py:1128]: xref-ing @pentschev's earlier comment https://github.com/dask/dask/pull/7321#discussion_r587902321

Comment (jrbourbeau) [reply dask/utils.py:1128]: The duck arrays I'm aware of (i.e. NumPy, sparse, cupy, pint, unyt) all will result in `is_duck_array` being `True`. That said, I can't be 100% certain that this covers all array-like libraries that are used with Dask today. Perhaps this is a good opportunity for us to specify the conditions for which an array library will work with Dask Array. Xarray has a nice documentation section on this http://xarray.pydata.org/en/stable/internals.html#internals-duck-arrays topic

**Comment by jrbourbeau** [comment]
If there are no further comments I'll propose we include this in the release tomorrow. @jakirkham @pentschev could I ask one of you to run the `cupy` tests with the changes here to ensure they pass?

**Comment by pentschev** [comment]
> If there are no further comments I'll propose we include this in the release tomorrow. @jakirkham @pentschev could I ask one of you to run the `cupy` tests with the changes here to ensure they pass?

I agree with changes here and tested it locally, the tests pass with the exception of those I mentioned in https://github.com/dask/dask/issues/7324 . I also applied the patch to a start just after #6738 and by doing so all tests would pass, so from the CuPy side I thing we're good.

Thanks @jrbourbeau for working on this! 😄

**Review by pentschev — APPROVED** [review]

**Comment by shoyer** [comment]
Thanks for working on this -- yes, this is important to keep Xarray working with dask :)

My only thought is that the name `is_arraylike()` is [not quite appropriate](https://numpy.org/neps/nep-0022-ndarray-duck-typing-overview.html#terminology), given that NumPy uses "array like" for a different meaning. I would suggest using the name `is_duck_array()` instead.

**Comment by pentschev** [comment]
> My only thought is that the name `is_arraylike()` is [not quite appropriate](https://numpy.org/neps/nep-0022-ndarray-duck-typing-overview.html#terminology), given that NumPy uses "array like" for a different meaning. I would suggest using the name `is_duck_array()` instead.

I do not oppose doing that, and semantically agree with it. I'm only wondering if renaming that may cause confusion with people, we also have other utility function such as `is_dataframe_like` and `is_series_like` with similar meaning, but perhaps renaming and adding that information to docstrings would suffice?

**Comment by jakirkham** [comment]
Given this is a bug fix we are trying to get in before a release and I don’t think this function is intended for usage outside Dask, maybe we can punt on function renaming and track that in a new issue?

**Review by jakirkham — APPROVED** [review]

**Comment by jrbourbeau** [comment]
> maybe we can punt on function renaming and track that in a new issue?

+1. @shoyer thanks for bringing this terminology issue up -- I agree there's a mismatch between the `is_arraylike` method and the term "array-like". Let's discuss in a follow-up issue

**Comment by jrbourbeau** [comment]
Thanks all for reviewing! Your feedback was very useful

**Comment by jakirkham** [comment]
Thanks James! 😄

**Comment by shoyer** [comment]
Thanks! I agree, this might be clarified well just by updating the
docstring for this helper function.

**Comment by jakirkham** [comment]
Filed issue ( https://github.com/dask/dask/issues/7327 ) on that point. Also noted updating the docstring. Please feel free to add anything else there 🙂

**Pull request merged**